---
title: "Current Smoking Prediction"
output: html_document
---

Prediction from deep learning model for current smoking generated by Zifeng Wang and transferred December 2019.

Starting datafiles:
smoking_prediction.tsv
isoform_estimation.tsv

```{r setup,include=F}
library(data.table)
library(vioplot)
library(knitr)
library(psych)
library(beeswarm)
library(pROC)
library(OptimalCutpoints)

```
```{r tab1,echo=F}
knitr::opts_chunk$set(echo = TRUE)
#m<-fread("/Users/petercastaldi/Desktop/COPDGene/RNASEQ/MasterFiles/master.file.freeze3.txt")
phe<-fread("/Users/petercastaldi/Desktop/COPDGene/Visit2/COPDGene_P1P2_All_Visit_25mar19.txt",data.table=F)
cbc<-fread("/Users/petercastaldi/Desktop/COPDGene/Visit2/COPDGene_P1P2_All_Visit_28feb19_CBCQC.txt",data.table=F)
phe2<-phe[phe$visitnum==2,]
cat("\nDimension of COPD Visit 2 phenotypes: ",dim(phe2))

d1<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/dataset_split.tsv")
d2<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/COPDGeneTest_actual_ids.csv")
allsubj<-c(d1$SubjectID,d2$actual_id)
length(unique(allsubj))

```


```{r select, echo=F}
predid<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/COPDGeneTest_actual_ids.csv")
exonbase<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/final_results/exon_base.tsv",data.table=F)
exonbase$sid<-predid$actual_id
names(exonbase)[1]<-"ebpred"

exon_gml<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal/exon_map_True_exon2gene_fs_False_256_128_64_final.tsv",data.table=F)
exon_gml$sid<-predid$actual_id
names(exon_gml)[1]<-"exon_gml"

exon_gml_fsl<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal/exon_map_True_exon2gene_fs_True_256_128_64_final.tsv",data.table=F)
exon_gml_fsl$sid<-predid$actual_id
names(exon_gml_fsl)[1]<-"exon_gml_fsl"

exon_iso<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal/exon+transcript_map_False_fs_False__final.tsv",data.table=F)
exon_iso$sid<-predid$actual_id
names(exon_iso)[1]<-"exon_iso"

all<-merge(phe2[,c("sid","SmokCigNow")],exonbase,by="sid")
all<-merge(all,exon_gml,by="sid")
all<-merge(all,exon_gml_fsl,by="sid")
all<-merge(all,exon_iso,by="sid")

roc_exonbase<-roc(SmokCigNow~ebpred,data=all,smooth=FALSE)
roc_exonbase

roc_exon_gml<-roc(SmokCigNow~exon_gml,data=all,smooth=FALSE)
roc_exon_gml

roc_exon_gml_fs<-roc(SmokCigNow~exon_gml_fsl,data=all,smooth=FALSE)
roc_exon_gml_fs

roc_exon_iso<-roc(SmokCigNow~exon_iso,data=all,smooth=FALSE)
roc_exon_iso

# accuracy<-sapply(all[3:6],function(x) )

# roc.test(roc_exonbase,roc_exongtffs,method="delong")
# 
# roc_exonbase<-roc(SmokCigNow~ebpred,data=all,smooth=TRUE)
# 
# roc_exongtffs<-roc(SmokCigNow~gtffspred,data=all,smooth=TRUE)
# 
# png("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/PJC/FIGURES/ROC.png")
# plot(roc_exonbase,cex.lab=1.4,cex.axis=1.4)
# plot(roc_exongtffs,add=TRUE,col="red",print.auc=FALSE)
# legend("bottomright",c("exon base","E-I map + FS"),pch=19,col=c("black","red"),xjust=1,cex=1.4)
# x<-dev.off()
# 
# 
# hist(csm$`Smoking Probability`,main="Distribution of Predicted Values, All Data")
# 
# vioplot(csm$`Smoking Probability`[csm$SmokCigNow==0],csm$`Smoking Probability`[csm$SmokCigNow==1],names=c("Former","Current"),main="Predictions by Current Smoking Status, Train Data")
# 
# vioplot(csm$SmokingPrediction[csm$IsInTestSet==1 & csm$SmokCigNow==0],csm$SmokingPrediction[csm$IsInTestSet==1 & csm$SmokCigNow==1],names=c("Former","Current"),main="Predictions by Current Smoking Status, Test Data")

```

## Beineke model comparisons, old data

```{r beineke,echo=F}

exon_gml_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene/exon_map_True_exon2gene_fs_False__final.tsv",data.table=F)
exon_gml_5$sid<-predid$actual_id
names(exon_gml_5)[1]<-"exon_gml_5"

exon_gml_fsl_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene/exon_map_True_exon2gene_fs_True__final.tsv",data.table=F)
exon_gml_fsl_5$sid<-predid$actual_id
names(exon_gml_fsl_5)[1]<-"exon_gml_fsl_5"

all<-merge(all,exon_gml_5,by="sid")
all<-merge(all,exon_gml_fsl_5,by="sid")

roc_exon_gml_5<-roc(SmokCigNow~exon_gml_5,data=all,smooth=FALSE)
roc_exon_gml_5

roc_exon_gml_fsl_5<-roc(SmokCigNow~exon_gml_fsl_5,data=all,smooth=FALSE)
roc_exon_gml_fsl_5



```
Look at characteristics of misclassified subjects. Use threshold of 0.5 to identify misclassification.

## New data, normalized

```{r newdatanormfull,echo=F}

exon_nn_base<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_new_data_normalized/exon_map_False_fs_False_256_128_64_final.tsv",data.table=F)
exon_nn_base$sid<-predid$actual_id
names(exon_nn_base)[1]<-"eb_nn_pred"

exon_nn_iml<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_new_data_normalized/exon_map_True_exon2transcript_fs_False_256_128_64_final.tsv",data.table=F)
exon_nn_iml$sid<-predid$actual_id
names(exon_nn_iml)[1]<-"exon_nn_iml"

exon_nn_iml_fsl<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_new_data_normalized/exon_map_True_exon2transcript_fs_True_256_128_64_final.tsv",data.table=F)
exon_nn_iml_fsl$sid<-predid$actual_id
names(exon_nn_iml_fsl)[1]<-"exon_nn_iml_fsl"

all_nn<-merge(phe2[,c("sid","SmokCigNow")],exon_nn_base,by="sid")
all_nn<-merge(all_nn,exon_nn_iml,by="sid")
all_nn<-merge(all_nn,exon_nn_iml_fsl,by="sid")

roc_exon_nn_base<-roc(SmokCigNow~eb_nn_pred,data=all_nn,smooth=FALSE)
roc_exon_nn_base

roc_exon_nn_iml<-roc(SmokCigNow~exon_nn_iml,data=all_nn,smooth=FALSE)
roc_exon_nn_iml

roc_exon_nn_iml_fs<-roc(SmokCigNow~exon_nn_iml_fsl,data=all_nn,smooth=FALSE)
roc_exon_nn_iml_fs


```

# Beineke model, normalized data

```{r beinekenew,echo=F}
gene_nn_base_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_normalized/gene_map_False_fs_False__final.tsv",data.table=F)
gene_nn_base_5$sid<-predid$actual_id
names(gene_nn_base_5)[1]<-"gene_nn_base_5"

exon_nn_base_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_normalized/exon_map_False_fs_False__final.tsv",data.table=F)
exon_nn_base_5$sid<-predid$actual_id
names(exon_nn_base_5)[1]<-"exon_nn_base_5"

iso_nn_base_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_normalized/transcript_map_False_fs_False__final.tsv",data.table=F)
iso_nn_base_5$sid<-predid$actual_id
names(iso_nn_base_5)[1]<-"iso_nn_base_5"

exon_nn_iml_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_normalized/exon_map_True_exon2transcript_fs_False__final.tsv",data.table=F)
exon_nn_iml_5$sid<-predid$actual_id
names(exon_nn_iml_5)[1]<-"exon_nn_iml_5"

exon_nn_iml_fsl_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_normalized/exon_map_True_exon2transcript_fs_True__final.tsv",data.table=F)
exon_nn_iml_fsl_5$sid<-predid$actual_id
names(exon_nn_iml_fsl_5)[1]<-"exon_nn_iml_fsl_5"

all_nn_5<-merge(phe2[,c("sid","SmokCigNow")],gene_nn_base_5,by="sid")
all_nn_5<-merge(all_nn_5,exon_nn_base_5,by="sid")
all_nn_5<-merge(all_nn_5,iso_nn_base_5,by="sid")
all_nn_5<-merge(all_nn_5,exon_nn_iml_5,by="sid")
all_nn_5<-merge(all_nn_5,exon_nn_iml_fsl_5,by="sid")

roc_gene_nn_base_5<-roc(SmokCigNow~gene_nn_base_5,data=all_nn_5,smooth=FALSE)
roc_gene_nn_base_5

roc_iso_nn_base_5<-roc(SmokCigNow~iso_nn_base_5,data=all_nn_5,smooth=FALSE)
roc_iso_nn_base_5

roc_exon_nn_base_5<-roc(SmokCigNow~exon_nn_base_5,data=all_nn_5,smooth=FALSE)
roc_exon_nn_base_5

roc_exon_nn_iml_5<-roc(SmokCigNow~exon_nn_iml_5,data=all_nn_5,smooth=FALSE)
roc_exon_nn_iml_5

roc_exon_nn_iml_fsl_5<-roc(SmokCigNow~exon_nn_iml_fsl_5,data=all_nn_5,smooth=FALSE)
roc_exon_nn_iml_fsl_5

```

# New data, normalized and covariate adjusted


```{r newdatanormcovfull,echo=F}

exon_nn_cov_base<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_new_data_covAdjusted/exon_map_False_fs_False_256_128_64_final.tsv",data.table=F)
exon_nn_cov_base$sid<-predid$actual_id
names(exon_nn_cov_base)[1]<-"eb_nn_cov_pred"

exon_nn_cov_iml<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_new_data_covAdjusted/exon_map_True_exon2transcript_fs_False_256_128_64_final.tsv",data.table=F)
exon_nn_cov_iml$sid<-predid$actual_id
names(exon_nn_cov_iml)[1]<-"exon_nn_cov_iml"

exon_nn_cov_iml_fsl<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_new_data_covAdjusted/exon_map_True_exon2transcript_fs_True_256_128_64_final.tsv",data.table=F)
exon_nn_cov_iml_fsl$sid<-predid$actual_id
names(exon_nn_cov_iml_fsl)[1]<-"exon_nn_cov_iml_fsl"

all_nn_cov<-merge(phe2[,c("sid","SmokCigNow")],exon_nn_cov_base,by="sid")
all_nn_cov<-merge(all_nn_cov,exon_nn_cov_iml,by="sid")
all_nn_cov<-merge(all_nn_cov,exon_nn_cov_iml_fsl,by="sid")

roc_exon_nn_cov_base<-roc(SmokCigNow~eb_nn_cov_pred,data=all_nn_cov,smooth=FALSE)
roc_exon_nn_cov_base

roc_exon_nn_cov_iml<-roc(SmokCigNow~exon_nn_cov_iml,data=all_nn_cov,smooth=FALSE)
roc_exon_nn_cov_iml

roc_exon_nn_cov_iml_fsl<-roc(SmokCigNow~exon_nn_cov_iml_fsl,data=all_nn_cov,smooth=FALSE)
roc_exon_nn_cov_iml_fsl


```

# Beineke model, normalized and covariate adjusted data

```{r beinekenewcov,echo=F}
gene_nn_cov_base_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_covAdjusted/gene_map_False_fs_False__final.tsv",data.table=F)
gene_nn_cov_base_5$sid<-predid$actual_id
names(gene_nn_cov_base_5)[1]<-"gene_nn_cov_base_5"

exon_nn_cov_base_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_covAdjusted/exon_map_False_fs_False__final.tsv",data.table=F)
exon_nn_cov_base_5$sid<-predid$actual_id
names(exon_nn_cov_base_5)[1]<-"exon_nn_cov_base_5"

iso_nn_cov_base_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_covAdjusted/transcript_map_False_fs_False__final.tsv",data.table=F)
iso_nn_cov_base_5$sid<-predid$actual_id
names(iso_nn_cov_base_5)[1]<-"iso_nn_cov_base_5"

exon_nn_cov_iml_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_covAdjusted/exon_map_True_exon2transcript_fs_False__final.tsv",data.table=F)
exon_nn_cov_iml_5$sid<-predid$actual_id
names(exon_nn_cov_iml_5)[1]<-"exon_nn_cov_iml_5"

exon_nn_cov_iml_fsl_5<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5_gene_new_data_covAdjusted/exon_map_True_exon2transcript_fs_True__final.tsv",data.table=F)
exon_nn_cov_iml_fsl_5$sid<-predid$actual_id
names(exon_nn_cov_iml_fsl_5)[1]<-"exon_nn_cov_iml_fsl_5"

all_nn_cov_5<-merge(phe2[,c("sid","SmokCigNow")],gene_nn_cov_base_5,by="sid")
all_nn_cov_5<-merge(all_nn_cov_5,exon_nn_cov_base_5,by="sid")
all_nn_cov_5<-merge(all_nn_cov_5,iso_nn_cov_base_5,by="sid")
all_nn_cov_5<-merge(all_nn_cov_5,exon_nn_cov_iml_5,by="sid")
all_nn_cov_5<-merge(all_nn_cov_5,exon_nn_cov_iml_fsl_5,by="sid")

roc_gene_nn_cov_base_5<-roc(SmokCigNow~gene_nn_cov_base_5,data=all_nn_cov_5,smooth=FALSE)
roc_gene_nn_cov_base_5

roc_iso_nn_cov_base_5<-roc(SmokCigNow~iso_nn_cov_base_5,data=all_nn_cov_5,smooth=FALSE)
roc_iso_nn_cov_base_5

roc_exon_nn_cov_base_5<-roc(SmokCigNow~exon_nn_cov_base_5,data=all_nn_cov_5,smooth=FALSE)
roc_exon_nn_cov_base_5

roc_exon_nn_cov_iml_5<-roc(SmokCigNow~exon_nn_cov_iml_5,data=all_nn_cov_5,smooth=FALSE)
roc_exon_nn_cov_iml_5

roc_exon_nn_cov_iml_fsl_5<-roc(SmokCigNow~exon_nn_cov_iml_fsl_5,data=all_nn_cov_5,smooth=FALSE)
roc_exon_nn_cov_iml_fsl_5

```

## Comparing  performance with and without removal of Beineke 5 genes and genes correlated with those 5 at > 0.4

```{r b5cor,echo=F}

allg<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5genesandcor_removed/1079gene.tsv",data.table=F)
allg$sid<-predid$actual_id
names(allg)[1]<-"all_pred"

subg<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/rebuttal_5genesandcor_removed/1020gene.tsv",data.table=F)
subg$sid<-predid$actual_id
names(subg)[1]<-"sub_pred"

all_b5c<-merge(phe2[,c("sid","SmokCigNow")],allg,by="sid")
all_b5c<-merge(all_b5c,subg,by="sid")

roc_allg<-roc(SmokCigNow~all_pred,data=all_b5c,smooth=FALSE)
roc_allg

roc_subg<-roc(SmokCigNow~sub_pred,data=all_b5c,smooth=FALSE)
roc_subg

```

##  Calculate classification accuracies using a threshold of 0.5
```{r accuracy,echo=F}
alllist<-list() 
for(a in names(all)[3:8]){
  alllist[[a]]<-prop.table(table(all$SmokCigNow==ifelse(all[,a]>0.5,1,0)))[2]
}
alllist


normlist<-list()
for(a in names(all_nn)[3:5]){
  normlist[[a]]<-prop.table(table(all_nn$SmokCigNow==ifelse(all_nn[,a]>0.5,1,0)))[2]
}
normlist

normlist_5<-list()
for(a in names(all_nn_5)[3:7]){
  normlist_5[[a]]<-prop.table(table(all_nn_5$SmokCigNow==ifelse(all_nn_5[,a]>0.5,1,0)))[2]
}
normlist_5

covlist<-list()
for(a in names(all_nn_cov)[3:5]){
  covlist[[a]]<-prop.table(table(all_nn_cov$SmokCigNow==ifelse(all_nn_cov[,a]>0.5,1,0)))[2]
}
covlist

covlist_5<-list()
for(a in names(all_nn_cov_5)[3:7]){
  covlist_5[[a]]<-prop.table(table(all_nn_cov_5$SmokCigNow==ifelse(all_nn_cov_5[,a]>0.5,1,0)))[2]
}
covlist_5

allb5clist<-list() 
for(a in names(all_b5c)[3:4]){
  allb5clist[[a]]<-prop.table(table(all_b5c$SmokCigNow==ifelse(all_b5c[,a]>0.5,1,0)))[2]
}
allb5clist

```

## Importance scores

```{r importance,echo=F}

imp<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/exon_tx_gene_mapping.csv",data.table=F)

hist(imp$saliency_importance,main="Saliency Importance")
hist(imp$fsl_importance,main="FSL Importance")

plot(imp$fsl_importance[imp$fsl_importance!=0 & imp$saliency_importance!=0],log(imp$saliency_importance[imp$fsl_importance!=0 & imp$saliency_importance!=0]),main="Non-zero importance scores")

table(imp$fsl_importance==0)
table(imp$saliency_importance==0)

imp_nonzero<-imp[imp$saliency_importance!=0 | imp$fsl_importance!=0,]
cat("Dimension of matrix with nonzero importance scores")
dim(imp_nonzero)

cat("\nNumber of genes with non-zero importance scores: \nFSL:",length(unique(imp$gene_id[imp$fsl_importance>0])),"\nSaliency: ",length(unique(imp$gene_id[imp$saliency_importance>0])))



```

## Test difference between ROCs in validation data

```{r roctest,echo=F}
predid<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/COPDGeneTest_actual_ids.csv")
exon<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/final_results/exon_base.tsv")
exonmap<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/final_results/Exon_Isoform_Map_GTF.tsv")
exonmapfs<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/final_results/Exon_Isoform_Map_GTF_FS.tsv")

exon$sid<-predid
exon<-merge(exon,phe2[,c("sid","SmokCigNow")],by="sid")

exonmap$sid<-predid
exonmap<-merge(exonmap,phe2[,c("sid","SmokCigNow")],by="sid")

exonmapfs$sid<-predid
exonmapfs<-merge(exonmapfs,phe2[,c("sid","SmokCigNow")],by="sid")

library(pROC)
exon_roc<-roc(exon$SmokCigNow,exon$`Smoking Probability`,plot=TRUE)
exonmap_roc<-roc(exonmap$SmokCigNow,exonmap$`Smoking Probability`,plot=TRUE)
exonmapfs_roc<-roc(exonmapfs$SmokCigNow,exonmapfs$`Smoking Probability`,plot=TRUE)

roc.test(exon_roc,exonmap_roc)
roc.test(exonmap_roc,exonmapfs_roc)
roc.test(exon_roc,exonmapfs_roc)
roc.test(exon_roc,roc_exon_gml_fs)

cat("\nTest change in performance from exon IML-FS model \ntrained in raw data and covariate  adjusted data:\n")
roc.test(exonmapfs_roc,roc_exon_nn_cov_iml_fsl)

```

## TOPGO enrichment analysis - conducted by Zhonghui. These results take importance scores for the feature selection layer and another metric called a saliency metric, and then we compare the genes  in the top 20% of importance to all of the other genes in the ~1000 gene set from Huan et al.

```{r toppgo.echo=F}
fsl_go<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/topgo/GOenrichDeepLearning_fsl_top20pct.tsv")

sal_go<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/topgo/GOenrichDeepLearning_saliency_top20pct.tsv")

imp<-fread("/Users/petercastaldi/Dropbox (Partners HealthCare)/Research/Others/Zifeng/prediction_data/rebuttal_test_prediction/topgo/exon_tx_gene_mapping.csv")

table(imp$saliency_importance>0)
length(unique(imp$gene[imp$saliency_importance>0]))

library(ggplot2)
ggplot(data.frame(imp$saliency_importance), aes(imp$saliency_importance)) +               # Histogram with log10 axis
  geom_histogram(bins = 100) +
  scale_x_log10()

quantile(imp$saliency_importance)
png()
hist(log(imp$saliency_importance[imp$saliency_importance>0]),breaks="scott")

cat("\nNumber of GO pathways at p<0.05 from FSL: ",length(unique(fsl_go$GO.ID[fsl_go$weight01Fisher<0.05])))

cat("\nNumber of GO pathways at p<0.05 from Saliency: ",length(unique(sal_go$GO.ID[sal_go$weight01Fisher<0.05])))

shared<-intersect(unique(fsl_go$GO.ID[fsl_go$weight01Fisher<0.05]),unique(sal_go$GO.ID[sal_go$weight01Fisher<0.05]))

cat("\nGO pathways present in both analyses: ", shared)

pm<-merge(fsl_go,sal_go,by="GO.ID",suffixes = c(".fsl",".sal"))

kable(pm[pm$GO.ID %in% shared,c(1:5,8,12,13,16)])

length(intersect(unique(fsl_go$GO.ID[fsl_go$weight01Fisher<0.05]),unique(sal_go$GO.ID[sal_go$weight01Fisher<0.05])))

pm[pm$weight01Fisher.sal<0.005,c(1:5,8,12,13,16)]

```

